magellan.1
.nh
.TH OpenCHAMI Magellan
.PP
The \fBmagellan\fR CLI tool is a Redfish-based, board management controller (BMC) discovery tool designed to scan networks and is written in Go. The tool collects information from BMC nodes using the provided Redfish RESTful API with 
\[la]https://github.com/stmcginnis/gofish\[ra] and loads the queried data into an SMD
\[la]https://github.com/OpenCHAMI/smd/tree/master\[ra] instance. The tool strives to be more flexible by implementing multiple methods of discovery to work for a wider range of systems (WIP) and is capable of using independently of other tools or services.

.PP
\fBNote: \fBmagellan\fR v0.1.0 is incompatible with SMD v2.15.3 and earlier.\fP

.SH Main Features
.PP
The \fBmagellan\fR tool comes packed with a handleful of features for doing discovery, such as:

.RS
.IP \(bu 2
Simple network scanning
.IP \(bu 2
Redfish-based inventory collection
.IP \(bu 2
Redfish-based firmware updating
.IP \(bu 2
Integration with OpenCHAMI SMD
.IP \(bu 2
Write inventory data to JSON

.RE

.PP
See the TODO
\[la]#todo\[ra] section for a list of soon-ish goals planned.

.SH Getting Started
.PP
Build
\[la]#building\[ra] and run on bare metal
\[la]#running\-the\-tool\[ra] or run and test with Docker using the latest prebuilt image
\[la]#running\-with\-docker\[ra]\&. For quick testing, the repository integrates a Redfish emulator that can be ran by executing the \fBemulator/setup.sh\fR script or running \fBmake emulator\fR\&.

.SH Building the Executable
.PP
The \fBmagellan\fR tool can be built to run on bare metal. Install the required Go tools, clone the repo, and then build the binary in the root directory with the following:

.EX
git clone https://github.com/OpenCHAMI/magellan 
cd magellan
go mod tidy && go build
.EE

.PP
And that's it. The last line should find and download all of the required dependencies to build the project. Although other versions of Go may work, the project has been tested to work with versions v1.20 and later on MacOS and Linux.

.SS Building on Debian 12 (Bookworm)
.PP
Getting the \fBmagellan\fR tool to work with Go 1.21 on Debian 12 may require installing the \fBgolang-1.21\fR meta-package from \fBbookworm-backports\fR through \fBapt\fR along with GCC for comping the \fBgo-sqlite3\fR driver.

.EX
apt install gcc golang-1.21/bookworm-backport
.EE

.PP
The binary executable for the \fBgolang-1.21\fR executable can then be found using \fBdpkg\fR\&.

.EX
dpkg -L golang-1.21-go
.EE

.PP
Using the correct binary, set the \fBCGO_ENABLED\fR environment variable and build the executable with \fBcgo\fR enabled:

.EX
export GOBIN=/usr/bin/golang-1.21/bin/go 
go env -w CGO_ENABLED=1
go mod tidy && go build
.EE

.PP
This might take some time to complete initially because of the \fBgo-sqlite3\fR driver, but should be much faster for subsequent builds.

.SS Docker
.PP
The tool can also run using Docker. To build the Docker container, run \fBdocker build -t magellan:testing .\fR in the project's directory. This is useful if you to run \fBmagellan\fR on a different system through Docker desktop without having to install and build with Go (or if you can't do so for some reason). Prebuilt images
\[la]https://github.com/OpenCHAMI/magellan/pkgs/container/magellan\[ra] are available as well on \fBghcr\fR\&. Images can be pulled directly from the repository:

.EX
docker pull ghcr.io/openchami/magellan:latest
.EE

.PP
See the "Running with Docker"
\[la]#running\-with\-docker\[ra] section below about running with the Docker container.

.SH Usage
.PP
The sections below assume that the BMC nodes have an IP address available to query Redfish. Currently, \fBmagellan\fR does not support discovery with MAC addresses although that may change in the future.

.SS Checking for Redfish
.PP
Before using the tool, confirm that the identified node has Redfish with \fBcurl\fR\&. Assuming the IP address for the BMC node is \fB172.16.0.10\fR, we can send a request to see if it we get a response. You might need to pass the \fB-k\fR flag if the node uses TLS or point to the appropriate certificate.

.EX
curl -k https://172.16.0.10/redfish/v1 --cacert cacert.pem | jq
.EE

.PP
This should return a JSON response with general information. The output below has been truncated:

.EX
{
  "@odata.context": "/redfish/v1/$metadata#ServiceRoot.ServiceRoot",
  "@odata.etag": "W/\\"1715279084\\"",
  "@odata.id": "/redfish/v1/",
  "@odata.type": "#ServiceRoot.v1_5_2.ServiceRoot",
  "AccountService": {
    "@odata.id": "/redfish/v1/AccountService"
  },
  "CertificateService": {
    "@odata.id": "/redfish/v1/CertificateService"
  },
  "Chassis": {
    "@odata.id": "/redfish/v1/Chassis"
  },
  ...
}
.EE

.SS Running the Tool
.PP
There are three main commands to use with the tool: \fBscan\fR, \fBlist\fR, and \fBcollect\fR\&. To see all of the available commands, run \fBmagellan\fR with the \fBhelp\fR subcommand:

.EX
Redfish-based BMC discovery tool

Usage:
  magellan [flags]
  magellan [command]

Available Commands:
  collect     Collect system information by interrogating BMC node
  completion  Generate the autocompletion script for the specified shell
  crawl       Crawl a single BMC for inventory information
  help        Help about any command
  list        List information stored in cache from a scan
  login       Log in with identity provider for access token
  scan        Scan to discover BMC nodes on a network
  update      Update BMC node firmware

Flags:
      --access-token string   set the access token
      --cache string          set the scanning result cache path (default "/tmp/allend/magellan/assets.db")
      --concurrency int       set the number of concurrent processes (default -1)
  -c, --config string         set the config file path
  -d, --debug                 set to enable/disable debug messages
  -h, --help                  help for magellan
      --timeout int           set the timeout (default 5)
  -v, --verbose               set to enable/disable verbose output

Use "magellan [command] --help" for more information about a command.
.EE

.PP
To start a network scan for BMC nodes, use the \fBscan\fR command. If the port is not specified, \fBmagellan\fR will probe the common Redfish port 443 by default:

.EX
\&./magellan scan \\\&
    --subnet 172.16.0.0 \\
    --subnet-mask 255.255.255.0 \\
    --format json \\
    --cache data/assets.db \\
.EE

.PP
This will scan the \fB172.16.0.0\fR subnet returning the host and port that return a response and store the results in a local cache with at the \fBdata/assets.db\fR path. Additional flags can be set such as \fB--host\fR to add more hosts to scan not included on the subnet, \fB--timeout\fR to set how long to wait for a response from the BMC node, or \fB--concurrency\fR to set the number of requests to make concurrently. Setting the \fB--format=json\fR will format the output in JSON. Try using \fB\&./magellan help scan\fR for a complete set of options this subcommand. Alternatively, the same scan can be started using CIDR notation and with additional hosts:

.EX
\&./magellan scan https://10.0.0.100:5000 --subnet 172.16.0.0/24
.EE

.PP
Check the help for each subcommand for more examples for specifying arguments.

.PP
To inspect the cache, use the \fBlist\fR command. Make sure to point to the same database used before:

.EX
\&./magellan list --cache data/assets.db --format json
.EE

.PP
This will print a list of node info found and stored from the scan. Like the \fBscan\fR subcommand, the output format can be set using the \fB--format\fR flag.

.PP
Finally, set the \fBACCESS_TOKEN\fRrun the \fBcollect\fR command to query the node from cache and send the info to be stored into SMD:

.EX
\&./magellan collect \\\&
    --cache data/assets.db \\
    --timeout 5 \\
    --username $USERNAME \\
    --password $PASSWORD \\
    --host https://example.openchami.cluster:27779 \\
    --output logs/
    --cacert cacert.pem
.EE

.PP
This uses the info stored in cache to request information about each BMC node if possible. Like with the scan, the time to wait for a response can be set with the \fB--timeout\fR flag as well. This command also requires the \fB--user\fR and \fB--pass\fR flags to be set if access the Redfish service requires basic authentication. Additionally, it may be necessary to set the \fB--host\fR and \fB--port\fR flags for \fBmagellan\fR to find the SMD API (not the root API endpoint "/hsm/v2"). The output of the \fBcollect\fR can be saved by using the \fB--output\fR

.PP
Note: If the \fBcache\fR flag is not set, \fBmagellan\fR will use "/tmp/$USER/magellan.db" by default.

.SS Updating Firmware
.PP
The \fBmagellan\fR tool is capable of updating firmware with using the \fBupdate\fR subcommand via the Redfish API. This may sometimes necessary if some of the \fBcollect\fR output is missing or is not including what is expected. The subcommand expects there to be a running HTTP/HTTPS server running that has an accessible URL path to the firmware download. Specify the URL with the \fB--firmware-path\fR flag and the firmware type with the \fB--component\fR flag with all the other usual arguments like in the example below:

.EX
\&./magellan update 172.16.0.108:443 \\\&
  --username $USERNAME \\ 
  --password $PASSWORD \\
  --firmware-path http://172.16.0.255:8005/firmware/bios/image.RBU \\
  --component BIOS
.EE

.PP
Then, the update status can be viewed by including the \fB--status\fR flag along with the other usual arguments or with the \fBwatch\fR command:

.EX
\&./magellan update 172.16.0.110 --status --username $USERNAME --pass $PASSWORD | jq '.'
# ...or...
watch -n 1 "./magellan update 172.16.0.110 --status --username $USERNAME --password $PASSWORD | jq '.'"
.EE

.SS Getting an Access Token (WIP)
.PP
The \fBmagellan\fR tool has a \fBlogin\fR subcommand that works with the 
\[la]https://github.com/OpenCHAMI/opaal\[ra] service to obtain a token needed to access the SMD service. If the SMD instance requires authentication, set the \fBACCESS_TOKEN\fR environment variable to have \fBmagellan\fR include it in the header for HTTP requests to SMD.

.EX
# must have a running OPAAL instance
\&./magellan login --url https://opaal:4444/login

# ...complete login flow to get token
export ACCESS_TOKEN=eyJhbGciOiJIUzI1NiIs...
.EE

.PP
Alternatively, if you are running the OpenCHAMI quickstart in the deployment recipes
\[la]https://github.com/OpenCHAMI/deployment\-recipes\[ra], you can run the provided script to generate a token and set the environment variable that way.

.EX
quickstart_dir=path/to/deployment/recipes/quickstart
source $quickstart_dir/bash_functions.sh
export ACCESS_TOKEN=$(gen_access_token)
.EE

.SS Running with Docker
.PP
The \fBmagellan\fR tool can be ran in a Docker container after pulling the latest image:

.EX
docker pull ghcr.io/openchami/magellan:latest

.EE

.PP
Then, run either with the helper script found in \fBbin/magellan.sh\fR or the binary in the container:

.EX
docker run ghcr.io/openchami/magellan:latest /magellan.sh --scan "--subnet 172.16.0.0 --port 443 --timeout 3" --collect "--user admin --pass password --host http://vm01 --port 27779"
# ... or ..
docker ghcr.io/openhami/magellan:latest /magellan scan --subnet 172.16.0.0 --subnet-mask 255.255.255.0
.EE

.SH How It Works
.PP
At its core, \fBmagellan\fR is designed to do three basic things:

.RS
.IP "  1." 5
Scan for BMC nodes in cluster available on a network
.IP "  2." 5
Query information about each BMC node through Redfish API
.IP "  3." 5
Store queried information into a system management database

.RE

.PP
First, the tool performs a scan to find running services on a network. This is done by sending a raw TCP packet to all specified hosts (either IP or host name) and taking note which services respond. At this point, \fBmagellan\fR has no way of knowing whether this is a Redfish service or not, so another HTTP request is made to verify. Once the BMC responds with an OK status code, \fBmagellan\fR will store the necessary information in a local cache database to allow collecting more information about the node later. This allows for users to only have to scan their cluster once to find systems that are currently available and scannable.

.PP
Next, the tool queries information about the BMC node using \fBgofish\fR API functions, but requires access to BMC node found in the scanning step mentioned above to work. If the node requires basic authentication, a user name and password is required to be supplied as well. Once the BMC information is retrieved from each node, the info is aggregated and a HTTP request is made to a SMD instance to be stored. Optionally, the information can be written to disk for inspection and debugging purposes.

.PP
In summary, \fBmagellan\fR needs at minimum the following configured to work on each node:

.RS
.IP "  1." 5
Available Redfish service with its known host and port
.IP "  2." 5
A running instance of SMD service with its known host and port
.IP "  3." 5
Docker to pull and run containers or Go to build binaries

.RE

.SH TODO
.PP
See the issue list
\[la]https://github.com/OpenCHAMI/magellan/issues\[ra] for plans for \fBmagellan\fR\&. Here is a list of other features left to add, fix, or do (and some ideas!):

.RS
.IP \(bu 2
[X] Confirm loading different components into SMD
.IP \(bu 2
[X] Add ability to set subnet mask for scanning
.IP \(bu 2
[ ] Add ability to scan with other protocols like LLDP and SSDP
.IP \(bu 2
[X] Add more debugging messages with the \fB-v/--verbose\fR flag
.IP \(bu 2
[ ] Separate \fBcollect\fR subcommand with making request to endpoint
.IP \(bu 2
[X] Support logging in with \fBopaal\fR to get access token
.IP \(bu 2
[X] Support using CA certificates with HTTP requests to SMD
.IP \(bu 2
[ ] Add tests for the regressions and compatibility
.IP \(bu 2
[X] Clean up, remove unused, and tidy code (first round)

.RE

.SH Copyright
.PP
Copyright

.PP
© 2023 Triad National Security, LLC. All rights reserved. This program was produced under U.S. Government contract 89233218CNA000001 for Los Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC for the U.S. Department of Energy/National Nuclear Security Administration. All rights in the program are reserved by Triad National Security, LLC, and the U.S. Department of Energy/National Nuclear Security Administration. The Government is granted for itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide license in this material to reproduce, prepare derivative works, distribute copies to the public, perform publicly and display publicly, and to permit others to do so.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).



## [0.0.5] - 2023-11-02

### Added

 * Ability to update firmware
 * Refactored connection handling for faster scanning
 * Updated to reflect home at github.com/OpenCHAMI
 * Updated to reflect ghcr.io as container home

 ## [Unreleased]

## [0.0.1] - 2023-09-14

### Added

* Ability to scan subnets for devices
* Ability to store results in a database
* Ability to generate an inventory from walking Redfish commands
* Ability to send inventory information to SMD


internal/magellan.go
package magellan

import (
	"context"
	"crypto/x509"
	"encoding/json"
	"fmt"
	"os"
	"time"

	bmclib "github.com/bmc-toolbox/bmclib/v2"
	"github.com/jacobweinstock/registrar"
	_ "github.com/mattn/go-sqlite3"
	_ "github.com/stmcginnis/gofish"
)

const (
	IPMI_PORT    = 623
	SSH_PORT     = 22
	TLS_PORT     = 443
	REDFISH_PORT = 5000
)

type BMCProbeResult struct {
	Host     string `json:"host"`
	Port     int    `json:"port"`
	Protocol string `json:"protocol"`
	State    bool   `json:"state"`
}

// NOTE: ...params were getting too long...
type QueryParams struct {
	Host          string
	Port          int
	User          string
	Pass          string
	Drivers       []string
	Preferred		string
	Timeout       int
	WithSecureTLS bool
	CertPoolFile  string
	Verbose       bool
	IpmitoolPath string
}

func NewClient(l *Logger, q *QueryParams) (*bmclib.Client, error) {
	// NOTE: bmclib.NewClient(host, port, user, pass)
	// ...seems like the `port` params doesn't work like expected depending on interface

	// tr := &http.Transport{
	// 	TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
	// }
	// httpClient := http.Client{
	// 	Transport: tr,
	// }

	// init client
	clientOpts := []bmclib.Option{
		// bmclib.WithSecureTLS(),
		// bmclib.WithHTTPClient(&httpClient),
		// bmclib.WithLogger(),
		// bmclib.WithRedfishHTTPClient(&httpClient),
		bmclib.WithRedfishPort(fmt.Sprint(q.Port)),
		bmclib.WithRedfishUseBasicAuth(true),
		bmclib.WithIpmitoolPort(fmt.Sprint(IPMI_PORT)),
		bmclib.WithIpmitoolPath(q.IpmitoolPath),
	}

	// only work if valid cert is provided
	if q.WithSecureTLS {
		var pool *x509.CertPool
		if q.CertPoolFile != "" {
			pool = x509.NewCertPool()
			data, err := os.ReadFile(q.CertPoolFile)
			if err != nil {
				return nil, fmt.Errorf("could not read cert pool file: %v", err)
			}
			pool.AppendCertsFromPEM(data)
		}
		// a nil pool uses the system certs
		clientOpts = append(clientOpts, bmclib.WithSecureTLS(pool))
	}
	// url := fmt.Sprintf("https://%s:%s@%s", q.User, q.Pass, q.Host)
	url := ""
	// if q.WithSecureTLS {
	// url = "https://"
	// } else {
	// 	url = "http://"
	// }

	if q.User == "" && q.Pass == "" {
		url += fmt.Sprintf("%s:%s@%s", q.User, q.Pass, q.Host)
	} else {
		url += q.Host
	}

	client := bmclib.NewClient(url, q.User, q.Pass, clientOpts...)
	ds := registrar.Drivers{}
	for _, driver := range q.Drivers {
		ds = append(ds, client.Registry.Using(driver)...) // ipmi, gofish, redfish
	}
	client.Registry.Drivers = ds

	return client, nil
}

func QueryMetadata(client *bmclib.Client, l *Logger, q *QueryParams) ([]byte, error) {
	// client, err := NewClient(l, q)

	// open BMC session and update driver registry
	ctx, ctxCancel := context.WithTimeout(context.Background(), time.Second*time.Duration(q.Timeout))
	client.Registry.FilterForCompatible(ctx)
	err := client.Open(ctx)
	if err != nil {
		ctxCancel()
		return nil, fmt.Errorf("could not open BMC client: %v", err)
	}

	defer client.Close(ctx)

	metadata := client.GetMetadata()
	if err != nil {
		ctxCancel()
		return nil, fmt.Errorf("could not get metadata: %v", err)
	}

	// retrieve inventory data
	b, err := json.MarshalIndent(metadata, "", "    ")
	if err != nil {
		ctxCancel()
		return nil, fmt.Errorf("could not marshal JSON: %v", err)
	}

	if q.Verbose {
		fmt.Printf("metadata: %v\n", string(b))
	}
	ctxCancel()
	return []byte(b), nil
}

func QueryInventory(client *bmclib.Client, l *Logger, q *QueryParams) ([]byte, error) {
	// discover.ScanAndConnect(url, user, pass, clientOpts)

	// open BMC session and update driver registry
	ctx, ctxCancel := context.WithTimeout(context.Background(), time.Second*time.Duration(q.Timeout))
	client.Registry.FilterForCompatible(ctx)
	err := client.PreferProvider(q.Preferred).Open(ctx)
	if err != nil {
		ctxCancel()
		return nil, fmt.Errorf("could not open client: %v", err)
	}
	defer client.Close(ctx)

	inventory, err := client.Inventory(ctx)
	if err != nil {
		ctxCancel()
		return nil, fmt.Errorf("could not get inventory: %v", err)
	}

	// retrieve inventory data
	b, err := json.MarshalIndent(inventory, "", "    ")
	if err != nil {
		ctxCancel()
		return nil, fmt.Errorf("could not marshal JSON: %v", err)
	}

	if q.Verbose {
		fmt.Printf("inventory: %v\n", string(b))
	}
	ctxCancel()
	return []byte(b), nil
}

func QueryPowerState(client *bmclib.Client, l *Logger, q *QueryParams) ([]byte, error) {
	ctx, ctxCancel := context.WithTimeout(context.Background(), time.Second*time.Duration(q.Timeout))
	client.Registry.FilterForCompatible(ctx)
	err := client.PreferProvider(q.Preferred).Open(ctx)
	if err != nil {
		ctxCancel()
		return nil, fmt.Errorf("could not open client: %v", err)
	}
	defer client.Close(ctx)

	inventory, err := client.GetPowerState(ctx)
	if err != nil {
		ctxCancel()
		return nil, fmt.Errorf("could not get inventory: %v", err)
	}

	// retrieve inventory data
	b, err := json.MarshalIndent(inventory, "", "    ")
	if err != nil {
		ctxCancel()
		return nil, fmt.Errorf("could not marshal JSON: %v", err)
	}

	if q.Verbose {
		fmt.Printf("power state: %v\n", string(b))
	}
	ctxCancel()
	return []byte(b), nil

}

func QueryUsers(client *bmclib.Client, l *Logger, q *QueryParams) ([]byte, error) {
	// discover.ScanAndConnect(url, user, pass, clientOpts)
	// client, err := NewClient(l, q)
	// if err != nil {
	// 	return nil, fmt.Errorf("could not make query: %v", err)
	// }

	// open BMC session and update driver registry
	ctx, ctxCancel := context.WithTimeout(context.Background(), time.Second*time.Duration(q.Timeout))
	client.Registry.FilterForCompatible(ctx)
	err := client.Open(ctx)
	if err != nil {
		ctxCancel()
		return nil, fmt.Errorf("could not open BMC client: %v", err)
	}

	defer client.Close(ctx)

	users, err := client.ReadUsers(ctx)
	if err != nil {
		ctxCancel()
		return nil, fmt.Errorf("could not get users: %v", err)
	}

	// retrieve inventory data
	b, err := json.MarshalIndent(users, "", "    ")
	if err != nil {
		ctxCancel()
		return nil, fmt.Errorf("could not marshal JSON: %v", err)
	}

	// return b, nil
	ctxCancel()
	if q.Verbose {
		fmt.Printf("users: %v\n", string(b))
	}
	return []byte(b), nil
}

func QueryBios(client *bmclib.Client, l *Logger, q *QueryParams) ([]byte, error) {
	// client, err := NewClient(l, q)
	// if err != nil {
	// 	return nil, fmt.Errorf("could not make query: %v", err)
	// }
	b, err := makeRequest(client, client.GetBiosConfiguration, q.Timeout)
	if q.Verbose {
		fmt.Printf("bios: %v\n", string(b))
	}
	return b, err
}

func makeRequest[T interface{}](client *bmclib.Client, fn func(context.Context) (T, error), timeout int) ([]byte, error) {
	ctx, ctxCancel := context.WithTimeout(context.Background(), time.Second*time.Duration(timeout))
	client.Registry.FilterForCompatible(ctx)
	err := client.Open(ctx)
	if err != nil {
		ctxCancel()
		return nil, fmt.Errorf("could not open client: %v", err)
	}

	defer client.Close(ctx)

	response, err := fn(ctx)
	if err != nil {
		ctxCancel()
		return nil, fmt.Errorf("could not get response: %v", err)
	}

	ctxCancel()
	return makeJson(response)
}

func makeJson(object interface{}) ([]byte, error) {
	b, err := json.MarshalIndent(object, "", "    ")
	if err != nil {
		return nil, fmt.Errorf("could not marshal JSON: %v", err)
	}
	return []byte(b), nil
}
